server/src/main/java/org/elasticsearch/common/network/ThreadWatchdog.java

```diff
@@ -0,0 +1,266 @@
+            final var prevValue = getAndIncrement();
+            assert isIdle(prevValue) : "thread [" + trackedThread.getName() + "] was already active";
+        }
+
+        public void stopActivity() {
+            assert trackedThread == Thread.currentThread() : trackedThread.getName() + " vs " + Thread.currentThread().getName();
+            final var prevValue = getAndIncrement();
+            assert isIdle(prevValue) == false : "thread [" + trackedThread.getName() + "] was already idle";
+        }
+
+        boolean isIdleOrMakingProgress() {
+            final var value = get();
+            if (isIdle(value)) {
+                return true;
+            }
+            if (value == lastObservedValue) {
+                // no change since last check
+                return false;
+            } else {
+                // made progress since last check
+                lastObservedValue = value;
+                return true;
+            }
+        }
+
+        private static boolean isIdle(long value) {
+            // the parity of the value indicates the idle state: initially zero (idle), so active == odd
+            return (value & 1) == 0;
+        }
+    }
+
+    public void run(Settings settings, ThreadPool threadPool, Lifecycle lifecycle) {
+        new Checker(threadPool, NETWORK_THREAD_WATCHDOG_INTERVAL.get(settings), NETWORK_THREAD_WATCHDOG_QUIET_TIME.get(settings), lifecycle)
+            .run();
+    }
+
+    /**
+     * Action which runs itself periodically, calling {@link #getStuckThreadNames} to check for active threads that didn't make progress
+     * since the last call, and if it finds any then it dispatches {@link #threadDumper} to log the current hot threads.
+     */
+    private final class Checker extends AbstractRunnable {
+        private final ThreadPool threadPool;
+        private final TimeValue interval;
+        private final TimeValue quietTime;
+        private final Lifecycle lifecycle;
+
+        Checker(ThreadPool threadPool, TimeValue interval, TimeValue quietTime, Lifecycle lifecycle) {
+            this.threadPool = threadPool;
+            this.interval = interval;
+            this.quietTime = quietTime.compareTo(interval) <= 0 ? interval : quietTime;
+            this.lifecycle = lifecycle;
+            assert this.interval.millis() <= this.quietTime.millis();
+        }
+
+        @Override
+        protected void doRun() {
+            if (isRunning() == false) {
+                return;
+            }
+
+            boolean rescheduleImmediately = true;
+            try {
+                final var stuckThreadNames = getStuckThreadNames();
+                if (stuckThreadNames.isEmpty() == false) {
+                    logger.warn("the following threads are active but did not make progress since the last scan: {}", stuckThreadNames);
+                    rescheduleImmediately = false;
+                    threadPool.generic().execute(threadDumper);
```

Comment:
I wonder if we need the entire thread dump or we could log only the message class name to be able to track it?
This way it is required less analysis to start working on the cause.

Reply:
We already identify the message in the `InboundHandler` logs and experience shows that this isn't enough to show why it's slow sometimes (see e.g. `_msearch` or `can_match` messages). Also at the level at which we're working here there is no concept of "message class name", we haven't aggregated anything yet so we're dealing with plain network bytes only.